compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -1944,6 +1953,15 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
 
         // Insert a Shallow borrow of the prefixes of any fake borrows.
         for place in fake_borrows {
+            if let PlaceBase::Local(l) = scrutinee_base
+                && l != place.local
+            {
+                // The base of this place is a temporary created for deref patterns. We don't emit
+                // fake borrows for these as they are not initialized in all branches.
+                // FIXME(deref_patterns): is this sound?
```

Comment:
This might not work for or patterns with a guard.

Reply:
Hmm, I can sort of see what you mean but not enough to come up with an example. Do you have one in mind?

Reply:
Oh, here's something that compiles and runs but shouldn't:
```rust
fn mut_or_pat_guard() {
    let mut b = Box::new(false);
    match b {
        _ if {
            *b = true;
            false
        } =>
        {
            unreachable!()
        }
        deref!(true) => {}
        _ => unreachable!(),
    }
}
```
Fake borrows are shallow, right? So borrowing `b` doesn't stop me from accessing `*b`. I'm not sure how we can prevent that, maybe we could take real borrows for places that have a deref pattern?

Reply:
If we keep shallow borrows then we need to shallow borrow inside deref patterns, which likely means adding a new kind of `PlaceElem`. I'd rather not do that

Reply:
Well, the borrows still have to be removed before codegen because they could point to inactive variants. So we need a non-shallow fake borrow.

Reply:
What issue do you expect this to cause?

Reply:
By itself at least, it works. I added a test.